masq_lib/src/percentage.rs

```diff
@@ -80,178 +229,418 @@ impl Percentage {
+            LoosePercentage::new(percent as u32).of(base).unwrap()
+        })
+    }
+
+    #[test]
+    fn loose_percentage_end_to_end_test_for_standard_values_signed() {
+        let expected_values = (-100..=0).rev().collect::<Vec<i8>>();
+
+        test_end_to_end(-100, expected_values, |percent, base| {
+            LoosePercentage::new(percent as u32).of(base).unwrap()
+        })
+    }
+
+    const TEST_SET: [Case; 5] = [
+        Case {
+            requested_percent: 101,
+            examined_base_number: 10000,
+            expected_result: 10100,
+        },
+        Case {
+            requested_percent: 150,
+            examined_base_number: 900,
+            expected_result: 1350,
+        },
+        Case {
+            requested_percent: 999,
+            examined_base_number: 10,
+            expected_result: 100,
+        },
+        Case {
+            requested_percent: 1234567,
+            examined_base_number: 20,
+            expected_result: 12345 * 20 + (67 * 20 / 100),
+        },
+        Case {
+            requested_percent: u32::MAX,
+            examined_base_number: 1,
+            expected_result: (u32::MAX / 100) as i64 + 1,
+        },
+    ];
+
+    #[test]
+    fn loose_percentage_for_large_values_unsigned() {
+        TEST_SET.into_iter().for_each(|case| {
+            let result = LoosePercentage::new(case.requested_percent)
+                .of(case.examined_base_number)
+                .unwrap();
+            assert_eq!(
+                result, case.expected_result,
+                "Expected {} does not match actual {}. Percents {} of base {}.",
+                case.expected_result, result, case.requested_percent, case.examined_base_number
+            )
+        })
+    }
+
+    #[test]
+    fn loose_percentage_end_to_end_test_for_large_values_signed() {
+        TEST_SET
+            .into_iter()
+            .map(|mut case| {
+                case.examined_base_number *= -1;
+                case.expected_result *= -1;
+                case
+            })
+            .for_each(|case| {
+                let result = LoosePercentage::new(case.requested_percent)
+                    .of(case.examined_base_number)
+                    .unwrap();
+                assert_eq!(
+                    result, case.expected_result,
+                    "Expected {} does not match actual {}. Percents {} of base {}.",
+                    case.expected_result, result, case.requested_percent, case.examined_base_number
+                )
+            })
+    }
+
+    #[test]
+    fn loose_percentage_multiple_of_percent_hits_limit() {
+        let percents = (u8::MAX as u32 + 1) * 100;
+        let subject = LoosePercentage::new(percents);
+
+        let result: Result<u8, BaseTypeOverflow> = subject.of(1);
+
+        assert_eq!(result, Err(BaseTypeOverflow {}))
+    }
+
+    #[test]
+    fn loose_percentage_multiplying_input_number_hits_limit() {
```

Comment:
You'd like to rephrase this test name and in the test below such that `hits_limit` is right after `loose_percentage`.

A rephrase of this test name would be `loose_percentage_hits_limit_at_multiplication`

For the test below this it would be `loose_percentage_hits_limit_at_addition_from_remainder`.